lib/rack/mock.rb

```diff
@@ -231,9 +231,11 @@ def cookie(name)
 
     def parse_cookies_from_header
       cookies = Hash.new
-      if original_headers.has_key? 'Set-Cookie'
-        set_cookie_header = original_headers.fetch('Set-Cookie')
-        set_cookie_header.split("\n").each do |cookie|
+      hashed_headers = Rack::Utils::HeaderHash[original_headers]
+      if hashed_headers.has_key?("Set-Cookie")
+        values = (v = hashed_headers["Set-Cookie"]).respond_to?(:to_ary) ? v.to_ary : v.split("\n")
```

Comment:
I see, the behaviour here is a bit... messed up.

Reply:
Yes, `HeaderHash` is a bit... under-specified, compared to the reality of headers and their oddities, as I notice has been [discussed at some length](https://github.com/rack/rack/issues/1598).  I am in violent agreement that "considering HTTP headers as general metadata is generally a mistake, in my experience. There are too many things which have specific semantic behaviour which requires specific things to handle."

I'm a bit down on `HeaderHash` overall, after discovering it doesn't handle `[["set-cookie", "foo=bar"], ["Set-Cookie", "baz=wombat"]]` correctly (though that's something for another PR).  Looking at its history, the joining of arrays by newlines has always been on a bit of an "as-needed" basis, and I'm loathe to change the behaviour in `#fetch`, because it appears to be used heavily and garbage collection, etc.

I've just pushed a new version of the PR, with an extra commit to make `HeaderHash#fetch` case-insensitive.  I also tried a different approach to handling "could be a string, could be an array of strings, who knows?" in the cookie parsing.  It's at least *differently* ugly this time.  Let me know which approach you prefer; I'm equally unhappy with both of them.